### ISSUE TITLE
Does LSIO need to exist?! Does `object_store` already do everything we need? If not, can we extend `object_store` instead of creating LSIO?

### ISSUE THREAD

**Issue by JackKelly** [JackKelly/light-speed-io#27]
[`object_store`](https://docs.rs/object_store/latest/object_store/) is a mature, well-supported crate which already does a lot of what we need:

- Provides an async, common interface to local filesystems and cloud filesystems (gcp, aws, azure, HTTP, WebDAV).
- [`ObjectStore::get_ranges`](https://docs.rs/object_store/latest/object_store/trait.ObjectStore.html#method.get_ranges) reads multiple byte ranges for a given path and - I think - returns multiple buffers.
- [`ObjectStore::put_multipart`](https://docs.rs/object_store/latest/object_store/trait.ObjectStore.html#tymethod.put_multipart) already splits large uploads into parallel chunks

Some things that `object_store` doesn't (yet) appear to support:

- io_uring
- In general, what happens if the user tries to launch, say, 1 million in-flight I/O operations. Will `object_store` submit all those million operations to the underlying storage subsystem? Or will `object_store` submit, say, 64 requests. And keep the number of in-flight requests fairly constant at around 64 requests in-flight at any moment.
- Reading into an existing buffer (such as reading into an existing numpy array in memory)
- `get_multipart` (to _download_ large objects in parallel)

## Relevant links:

object_store issues:
* https://github.com/apache/arrow-rs/issues/4631 (where I explicitly asked about io_uring. TL;DR: The author would like to see some benchmarks!)
* https://github.com/apache/arrow-rs/issues/5240

Discussion on object_store PRs:
* https://github.com/apache/arrow-rs/pull/4677#discussion_r1291216153
* https://github.com/apache/arrow-rs/pull/5281#discussion_r1442762521

**Comment by JackKelly** [comment]
## What would "`io_uring` in `object_store`" look like?

### Options which _don't_ change `object_store`'s public API:

* `impl ObjectStore for IoUringLocal`
* `get*` methods would stat the file, create a buffer, then chain open-read-close in io_uring.
* I guess the `IoUringLocal` struct would own a single `io_uring` instance? (Is that possible? Or would each `Future` returned by `get*` have its own `io_uring` instance??)

What happens if the user naively tries to read one million files at once (by calling `get*` one million times)?:

* When running in `tokio`, _every_ call to `get*` will be run in `tokio::spawn_blocking`, which will launch a new thread. This is fine for a few hundred calls. But might not be fine for a few _million_ calls??
* (Although, maybe I could change that behaviour? Although, I'm not sure how)
* Whichever async runtime is used, it'll have to keep around one `Future` for every file.
* We'd have at least one system call per file.
* We can't re-use buffers.

What happens if the user takes responsibility for only having, say, a max of 64 operations in flight at any one time?

* We no longer need to worry about spawning too many tokio threads, or having too many `Future`s around.
* We'd still have to worry about having one system call per file. Which somewhat reduces the benefits of using io_uring!
* We can't re-use buffers.

### Options which _do_ change `object_store`'s public API:

Add these methods to the `ObjectStore` trait:
* `get_objects`: returns a `Stream` of buffers.
    * Provide a simple default implementation of `get_objects` which just calls `ObjectStore::get` in a loop.
* `get_into`: Reads data directly into a user-supplied buffer
    * Provide a simple default implementation which calls `get` and memcopies into the user-supplied buffer.

**Comment by JackKelly** [comment]
## Next steps

Quite rightly, the `object_store` maintainer would like to see some benchmarks. If `io_uring` isn't demonstrably faster than existing `object_store`, then what's the point!

I should order my new workstation! (Threadripper + PCIe 5 SSD)

Some thoughts about benchmarks:

* Workloads (all reading from a local SSD):
    * Read 1 million small files
    * Read 10,000 small _chunks_ from 100 files. (i.e. read 10,000 chunks from the first file, and 10,000 chunks from the second file, etc.)
    * Try the above with and without decompression (in parallel)
* Code to benchmark (against the above workloads):
    * LSIO using io_uring
    * `object_store` where we naively submit all 1,000,000 ops to `object_store` in one go
    * `object_store` where we are careful not to have more than 64 ops in-flight at any one time.

**Comment by JackKelly** [comment]
Actually, maybe the first step should be to compare `fio` (using io_uring) to `object_store`? (Although I don't think `fio` can decompress data?)

**Comment by JackKelly** [comment]
## Should I fork `object_store`? Or should LSIO be an extension crate for `object_store`?

### Advantages of forking `object_store` (with the ultimate aim being to merge a PR into `object_store`)

- Existing users of `object_store` would see performance benefits immediately, without any additional dependencies.

### Advantages of LSIO being an extension crate

- Doesn't burden `object_store` with io_uring code
- I _think_ that it should still be possible to implement everything I need. I can define a new `GetObjects` trait, with a `get_objects` method, with a default implementation that just calls `get` in a loop, and I can `impl GetObjects` for all existing `object_store` structs.
- Once we have benchmarks and a clearer idea of how to add `io_uring` to `object_store` then maybe we can consider merging the relevant code into `object_store`. It's not uncommon in the Rust community for new ideas to begin life as a separate crate, and then be merged into another crate.
- I can iterate faster

**Comment by JackKelly** [comment]
## Do we need a `get_objects(objects) -> Stream` method? Or will the existing `ObjectStore::get` suffice?

It'd be lovely to _not_ have to change `object_store`'s public API. But there are several questions that need answering. I'm not yet sure of the answers:

If we _don't_ have a `get_objects` method, then:

1. Can we efficiently submit operations to io_uring, and efficiently receive operations? Specifically: 
    - is there an elegant way to limit the number of SQEs in-flight to, say, 64? And to keep io_uring's SQ topped up?
    - when a CQE arrives, how do we "unblock" the associated `Future`?
 2. How would users elegantly interleave IO with compute. Data might arrive from IO in any order. As soon as a buffer is filled with data from IO, we want to schedule computation on rayon's threadpool. Maybe it's as simple as something like:

```rust
for file in files {
    let future: Future = store.get(file);
    let processed_data = tokio::spawn(move async || {
        let data = future.await;
        
        // We now have the data! So let's submit to Rayon's threadpool:
        // Adapted from https://ryhl.io/blog/async-what-is-blocking/#the-rayon-crate
        let (send, recv) = tokio::sync::oneshot::channel();
        rayon::spawn(move || {
            let result = process(data);
            let _ = send.send(result);
        });
        recv.await
    });
}
```

**Comment by JackKelly** [comment]
To save us from having to re-implement _every_ `ObjectStore` method for `IoUringLocal`, maybe the `IoUringLocal` struct can _contain_ a `LocalFileSystem`, and we can use the `delegate` crate to delegate method calls to the inner `LocalFileSystem`. That way, the `IoUringLocal` struct can be fully functional from day one, and we can slowly replace delegated functions with custom io_uring implementations. 

Also, we could have a `BatchObjectStore` trait with equivalents of `ObjectStore` methods, but which accept a vector of operations, and returns a stream. The trait would include default implementations which just call the single-op `ObjectStore` methods in a loop, like the code in the comment above

**Comment by JackKelly** [comment]
I think I've moved all the ideas from this issue into a bunch of separate tasks (each described by GitHub issues). I'll close this (very long) issue now, in preference of the new milestone: https://github.com/JackKelly/light-speed-io/milestone/1

**Comment by JackKelly** [comment]
Cross-referencing with relevant issues from the milestone:

- #29 
- #30 
- #31 
- #32 
- #33 
- #34 
- #35 
- #36 
- #37 
- #38
